File: src/bin/mt/apksign/data/DataSources.java
package bin.mt.apksign.data;

import bin.io.RandomAccessFile;

import java.io.IOException;

public class DataSources {

    public static DataSource fromFile(RandomAccessFile randomAccessFile, long start, long size) {
        return new FileDataSource(randomAccessFile, start, size);
    }

    public static DataSource fromData(byte[] data) {
        return fromData(data, 0, data.length);
    }

    public static DataSource fromData(byte[] data, int start, int size) {
        return new ByteArrayDataSource(data, start, size);
    }

    public static DataSource align(DataSource source, int align) {
        long size = source.size();
        int overCount = (int) (size % align);
        if (overCount == 0)
            return source;
        int fillCount = align - overCount;
        return link(source, fromData(new byte[fillCount]));
    }

    public static DataSource link(DataSource... sources) {
        return new ChainedDataSource(sources);
    }

    public static void reset(DataSource... sources) throws IOException {
        for (DataSource source : sources) {
            source.reset();
        }
    }

}

File: Android/app/src/main/java/bin/mt/apksign/data/DataSources.kt
package bin.mt.apksign.data

import bin.io.RandomAccessFile
import java.io.IOException

object DataSources {
    fun fromFile(randomAccessFile: RandomAccessFile, start: Long, size: Long): DataSource {
        return FileDataSource(randomAccessFile, start, size)
    }

    @JvmOverloads
    fun fromData(data: ByteArray, start: Int = 0, size: Int = data.size): DataSource {
        return ByteArrayDataSource(data, start, size)
    }

    fun align(source: DataSource, align: Int): DataSource {
        val size = source.size()
        val overCount = (size % align).toInt()
        if (overCount == 0) return source
        val fillCount = align - overCount
        return link(source, fromData(ByteArray(fillCount)))
    }

    fun link(vararg sources: DataSource): DataSource {
        return ChainedDataSource(*sources)
    }

    @Throws(IOException::class)
    fun reset(vararg sources: DataSource) {
        for (source in sources) {
            source.reset()
        }
    }
}